src/test/java/io/jenkins/plugins/analysis/core/quality/QualityGateTest.java

```diff
@@ -0,0 +1,230 @@
+    }
+
+
+    /**
+     * Test the test set defined by the parameters against all four thresholds (total, high, normal, low).
+     * The test method implements soft assertions. All assertions are executed even a assertion fails.
+     * @param builder function which sets the thresholds to test
+     * @param threshold to set
+     * @param totalWarningCount count of all warnings
+     * @param newWarningCount count of new warnings
+     * @param result expected result of the tests
+     */
+    private void testThreshold(Function<ThresholdSet, QualityGateBuilder> builder, final int threshold, final int totalWarningCount, final int newWarningCount, final Result result) {
+        QualityGate qualityGateTotal = builder.apply(new ThresholdSet(threshold, 0, 0, 0)).build();
+        QualityGate qualityGateHigh = builder.apply(new ThresholdSet(Integer.MAX_VALUE, threshold, 0, 0)).build();
+        QualityGate qualityGateNormal = builder.apply(new ThresholdSet(Integer.MAX_VALUE, 0, threshold, 0)).build();
+        QualityGate qualityGateLow = builder.apply(new ThresholdSet(Integer.MAX_VALUE, 0, 0, threshold)).build();
+
+        StaticAnalysisRun runTotal = mock(StaticAnalysisRun.class);
+        when(runTotal.getTotalSize()).thenReturn(totalWarningCount);
+        when(runTotal.getNewSize()).thenReturn(newWarningCount);
+
+        StaticAnalysisRun runHigh = mock(StaticAnalysisRun.class);
+        when(runHigh.getTotalHighPrioritySize()).thenReturn(totalWarningCount);
+        when(runHigh.getNewHighPrioritySize()).thenReturn(newWarningCount);
+
+        StaticAnalysisRun runNormal = mock(StaticAnalysisRun.class);
+        when(runNormal.getTotalNormalPrioritySize()).thenReturn(totalWarningCount);
+        when(runNormal.getNewNormalPrioritySize()).thenReturn(newWarningCount);
+
+        StaticAnalysisRun runLow = mock(StaticAnalysisRun.class);
+        when(runLow.getTotalLowPrioritySize()).thenReturn(totalWarningCount);
+        when(runLow.getNewLowPrioritySize()).thenReturn(newWarningCount);
+
+        SoftAssertions softly = new SoftAssertions();
+        softly.assertThat(qualityGateTotal.evaluate(runTotal))
+                .as("Threshold was <" + threshold + "> warning count was <" + totalWarningCount + ">")
+                .isEqualTo(result);
+        softly.assertThat(qualityGateHigh.evaluate(runHigh))
+                .as("High priority threshold was <" + threshold + "> warning count was <" + totalWarningCount + ">")
+                .isEqualTo(result);
+        softly.assertThat(qualityGateNormal.evaluate(runNormal))
+                .as("Normal priority threshold was <" + threshold + "> warning count was <" + totalWarningCount + ">")
+                .isEqualTo(result);
+        softly.assertThat(qualityGateLow.evaluate(runLow))
+                .as("Low priority threshold was <" + threshold + "> warning count was <" + totalWarningCount + ">")
+                .isEqualTo(result);
+        softly.assertAll();
+    }
+
+
+}
```

Comment:
Mir fehlt noch der Fall, dass ein Failure immer einen Unstable schlägt. D.h. mehrere Prüfungen sollten anschlagen, jeweils mit verschiedenem Resultat.

Reply:
Der Source Code ist sehr kompakt und entspricht dem DRY Prinzip. Insgesamt fällt es aber auf den ersten Blick schwer, die eigentliche Funktionalität zu erkennen. Vielleicht lässt sich das mit dem Punkt vorher kombinieren, dass man mal nur das API in Action sieht ohne die testThreshold Methode.

Reply:
Insgesamt prima! Ich würde diesen PR gerne mergen. Falls Sie daher noch Lust und Zeit auf die o.g. Punkte haben...

Reply:
Ich werde mich heute Abend um die bemängelten Punkte kümmern.

Reply:
Der Fall, dass Failure Unstable schlägt wird in den Tests, die auf den folgenden Namen hören getestet:
shouldBeFailedWhen(New)|(Total)FailedAnd[(New)|(Total)]UnstableThresholdIsEqualThanWarningCount
Hier wird der Failure und der Unstable Threshold jeweils mit den gleichen Werten beladen.

Reply:
Ah, scollen nach rechts hilft. Jetzt auf meinem Monitor (vorher Laptop) sehe ich den ganzen Quelltext ;-)